tests/src/main/java/net/neoforged/neoforge/debug/conditions/ConditionalCodecTest.java

```diff
@@ -0,0 +1,246 @@
+                      "i": 1,
+                      "s": "test"
+                    }""", write(SimpleRecord.CONDITIONS_CODEC, Optional.of(WithConditions.builder(record).addCondition(TrueCondition.INSTANCE).build())));
+        }
+
+        private record SimpleRecord(int i, String s) {
+            public static Codec<SimpleRecord> CODEC = RecordCodecBuilder.create(
+                    builder -> builder
+                            .group(
+                                    Codec.INT.fieldOf("i").forGetter(SimpleRecord::i),
+                                    Codec.STRING.fieldOf("s").forGetter(SimpleRecord::s))
+                            .apply(builder, SimpleRecord::new));
+            public static Codec<Optional<SimpleRecord>> CONDITIONAL_CODEC = ConditionalOps.createConditionalCodec(CODEC);
+            public static Codec<Optional<WithConditions<SimpleRecord>>> CONDITIONS_CODEC = ConditionalOps.createConditionalCodecWithConditions(CODEC);
+        }
+    }
+
+    private static class Primitive {
+        private static final Codec<Optional<Integer>> CONDITIONAL_INT = ConditionalOps.createConditionalCodec(Codec.INT);
+        private static final Codec<Optional<WithConditions<Integer>>> CONDITIONS_INT = ConditionalOps.createConditionalCodecWithConditions(Codec.INT);
+
+        public static void testReadNoConditions() {
+            JsonElement json = read("""
+                    1
+                    """);
+            int decoded = CONDITIONAL_INT.decode(JsonOps.INSTANCE, json).result().get().getFirst().get();
+            assertEquals(1, decoded);
+        }
+
+        /**
+         * Nesting a value without conditions is not allowed (why would anyone do this?).
+         */
+        public static void testReadNestedNoConditions() {
+            JsonElement json = read("""
+                    {
+                        "neoforge:value": 1
+                    }
+                    """);
+            assertErrored(CONDITIONAL_INT.decode(JsonOps.INSTANCE, json));
+        }
+
+        public static void testReadMatchingConditions() {
+            JsonElement json = read("""
+                    {
+                        "neoforge:conditions": [
+                            { "type": "neoforge:true" }
+                        ],
+                        "neoforge:value": 1
+                    }
+                    """);
+            int decoded = CONDITIONAL_INT.decode(JsonOps.INSTANCE, json).result().get().getFirst().get();
+            assertEquals(1, decoded);
+        }
+
+        public static void testReadFailingConditions() {
+            JsonElement json = read("""
+                    {
+                        "neoforge:conditions": [
+                            { "type": "neoforge:false" }
+                        ],
+                        "i": 1,
+                        "s": "test"
+                    }
+                    """);
+            // We expect a partial result with Optional.empty()
```

Comment:
It now expects (and tests for) a normal result with Optional.empty()

Reply:
Thanks!